web-api/src/personalization/stateless.rs

```diff
@@ -0,0 +1,369 @@
+    storage::{self, TagWeights},
+    Error,
+};
+
+#[derive(Deserialize)]
+pub(super) struct UnvalidatedHistoryEntry {
+    id: String,
+    #[serde(default)]
+    timestamp: Option<DateTime<Utc>>,
+}
+
+#[derive(Deserialize, PartialEq, Debug)]
+pub(super) struct HistoryEntry {
+    pub(super) id: DocumentId,
+    pub(super) timestamp: DateTime<Utc>,
+}
+
+/// Validates given history.
+///
+/// The history is expected to be ordered from oldest to
+/// newest entry, i.e. new entries are pushed to the end
+/// of the history vec.
+pub(super) fn validate_history(
+    history: Vec<UnvalidatedHistoryEntry>,
+    config: &PersonalizationConfig,
+    warnings: &mut Vec<Warning>,
+    time: DateTime<Utc>,
+    allow_empty_history: bool,
+) -> Result<Vec<HistoryEntry>, Error> {
+    if !allow_empty_history && history.is_empty() {
+        return Err(HistoryTooSmall.into());
+    }
+    let max_history_len = config.max_stateless_history_size;
+    if history.len() > max_history_len {
+        warnings.push(format!("history truncated, max length is {max_history_len}").into());
+    }
+    let mut most_recent_time = time;
+    let mut history = history
+        .into_iter()
+        .rev()
+        .take(max_history_len)
+        .map(|unchecked| {
+            let id = DocumentId::try_from(unchecked.id)?;
+            let timestamp = unchecked.timestamp.unwrap_or(most_recent_time);
+            if timestamp > most_recent_time {
+                warnings.push(format!("inconsistent history ordering around document {id}").into());
+            }
+            most_recent_time = timestamp;
+            Ok(HistoryEntry { id, timestamp })
+        })
+        .try_collect::<_, Vec<_>, Error>()?;
+
+    history.reverse();
+    Ok(history)
+}
+
+pub(super) async fn load_history(
+    storage: &impl storage::Document,
+    history: Vec<HistoryEntry>,
+) -> Result<Vec<LoadedHistoryEntry>, Error> {
+    let mut loaded =
+        storage::Document::get_interacted(storage, history.iter().map(|entry| &entry.id))
+            .await?
+            .into_iter()
+            .map(|document| (document.id, (document.embedding, document.tags)))
+            .collect::<HashMap<_, _>>();
+
+    Ok(history
+        .into_iter()
+        // filter ignores documents which don't exist in our database (i.e. have
+        // been deleted)
+        .filter_map(|HistoryEntry { id, timestamp }| {
+            loaded
+                .remove(&id)
+                .map(|(embedding, tags)| LoadedHistoryEntry {
+                    id,
+                    timestamp,
+                    embedding,
+                    tags,
+                })
+        })
+        .collect())
+}
+
+pub(super) struct LoadedHistoryEntry {
+    pub(super) id: DocumentId,
+    pub(super) timestamp: DateTime<Utc>,
+    pub(super) embedding: NormalizedEmbedding,
+    pub(super) tags: Vec<DocumentTag>,
+}
+
+/// Given an iterator over the history from _newest_ to oldest calculates user interests and tag weights.
```

Comment:
❓ shouldn't it be the opposite? if we have an iterator from newest to oldest history, then the shifts to the cois in `log_positive_user_reaction()` for newer interactions will be overwritten by shifts for older interactions.

Reply:
yes you are right, this was already wrong in the previous version :see_no_evil:

Reply:
turns out it was only a bug in the documentation :tada: